Cython/Compiler/Symtab.py

```diff
@@ -2301,17 +2329,15 @@ def get_refcounted_entries(self, include_weakref=False,
         return have_entries, (py_attrs, py_buffers, memoryview_slices)
 
     def declare_var(self, name, type, pos,
-                    cname = None, visibility = 'private',
-                    api = 0, in_pxd = 0, is_cdef = 0):
+                    cname=None, visibility='private',
+                    api=False, in_pxd=False, is_cdef=False, pytyping_modifiers=None):
         name = self.mangle_class_private_name(name)
 
-        if type.python_type_constructor_name == "typing.ClassVar":
-            is_cdef = 0
-            type = type.resolve()
-
-        if (type.python_type_constructor_name == "dataclasses.InitVar" and
-                'dataclasses.dataclass' not in self.directives):
-            error(pos, "Use of cython.dataclasses.InitVar does not make sense outside a dataclass")
+        if pytyping_modifiers:
+            if "typing.ClassVar" in pytyping_modifiers:
+                is_cdef = 0
```

Comment:
```suggestion
                is_cdef = 0
                if not type.is_pyobject:
                    type = type.equivalent_type or py_object_type
```

This is sufficient to get the annotation_typing and cdef_class_dataclasses tests to compile successfully on my PC. Both tests look to need minor tweaks to get them to pass after that (but that's stuff like the line numbers on warnings rather that real problems)

Reply:
Eventually, I would like to allow C types here. Not for the class variable but for its usage, type inference, etc.
However, it seems that that time isn't now.
I don't think requiring an object type here is a regression in any way. It's just a missing feature.
Let's see if it works. Do you think we should issue a warning here? If users request a C type, and Cython overrides it, I think they should know.

Reply:
I think a warning is probably a good idea.

Possibly an error for anything that genuinely can't be replaced by a Python object too?